src/plugins/visualizations/public/legacy/build_pipeline.ts

```diff
@@ -314,65 +283,122 @@ export const buildPipelineVisFunction: BuildPipelineVisFunction = {
-    }
+    const fontFn = style
+      ? buildExpression([buildExpressionFunction('font', { size: style.fontSize })])
+      : undefined;
+
+    const colorRangeFns = colorsRange.map((range: { from: number; to: number }) => {
+      return buildExpression([
+        buildExpressionFunction('range', {
+          from: range.from,
+          to: range.to,
+        }),
+      ]);
+    });
 
-    if (colorsRange) {
-      colorsRange.forEach((range: any) => {
-        expr += prepareValue('colorRange', `{range from=${range.from} to=${range.to}}`, true);
-      });
+    const visdimensionMetricFns = metrics.length
+      ? metrics.map((metric: SchemaConfig) => {
+          const fn = buildExpressionFunction('visdimension', { accessor: metric.accessor });
+          if (metric.format?.id) {
+            fn.addArgument('format', metric.format.id);
+          }
+          if (metric.format?.params) {
+            fn.addArgument('formatParams', JSON.stringify(metric.format.params));
+          }
+          return buildExpression([fn]);
+        })
+      : undefined;
+
+    const visdimensionBucketFn =
+      bucket && style
+        ? buildExpressionFunction('visdimension', { accessor: bucket.accessor })
+        : undefined;
+
+    if (visdimensionBucketFn && bucket.format) {
+      if (bucket.format?.id) {
+        visdimensionBucketFn.addArgument('format', bucket.format.id);
+      }
+      if (bucket.format?.params) {
+        visdimensionBucketFn.addArgument('formatParams', JSON.stringify(bucket.format.params));
+      }
     }
 
-    metrics.forEach((metric: SchemaConfig) => {
-      expr += prepareDimension('metric', metric);
-    });
+    const args = {
+      percentageMode,
+      colorSchema,
+      colorMode,
+      useRanges,
+      invertColors,
+      showLabels: labels?.show,
+      bgFill: style?.bgFill,
+      font: fontFn,
+      subText: style?.subText,
+      colorRange: colorRangeFns,
+      metric: visdimensionMetricFns ? visdimensionMetricFns : undefined,
+      bucket: visdimensionBucketFn ? buildExpression([visdimensionBucketFn]) : undefined,
+    };
 
-    return expr;
+    return buildExpressionFunction('metricvis', {
+      ...omit(args, i => i === null || typeof i === 'undefined'),
+    }).toString();
   },
   tagcloud: (params, schemas) => {
     const { scale, orientation, minFontSize, maxFontSize, showLabel } = params;
-    const { metric, bucket } = buildVisConfig.tagcloud(schemas);
-    let expr = `tagcloud metric={visdimension ${metric.accessor}} `;
-    expr += prepareValue('scale', scale);
-    expr += prepareValue('orientation', orientation);
-    expr += prepareValue('minFontSize', minFontSize);
-    expr += prepareValue('maxFontSize', maxFontSize);
-    expr += prepareValue('showLabel', showLabel);
-    expr += prepareDimension('bucket', bucket);
-
-    return expr;
+    const visConfig = buildVisConfig.tagcloud(schemas);
+
+    const visdimensionMetric = visConfig.metric
```

Comment:
metric needs to be preent for tagcloud, so i don't think we need to check for it